qiskit/optimization/algorithms/optimization_algorithm.py

```diff
@@ -113,17 +113,17 @@ def __init__(self, x: Optional[Any] = None, fval: Optional[Any] = None,
                  results: Optional[Any] = None,
                  status: OptimizationResultStatus = OptimizationResultStatus.SUCCESS,
                  variables: Optional[List[Variable]] = None) -> None:
-        self._x = x if x else []   # pylint: disable=invalid-name
-        self._variables = variables if variables else []
+        self._x = x if x is not None else []   # pylint: disable=invalid-name
+        self._variables = variables if variables is not None else []
         self._fval = fval
         self._results = results
         self._status = status
         self._variable_names = [variable.name for variable in self._variables]
-        self._variables_dict = dict(zip(self._variable_names, self._x))
+        self._variables_dict = dict(zip(self._variable_names or [], self._x or []))
 
     def __repr__(self):
         return 'optimal variables: [{}]\noptimal function value: {}\nstatus: {}' \
-            .format(','.join([str(x_) for x_ in self.x]), self.fval, self.status.name)
+            .format(','.join([str(x_) for x_ in self.x or []]), self.fval, self.status.name)
```

Comment:
The `self.x or []` has the same issue as the variable assignment in that self.x may be a valid value which as a bool is False hence again it should be an explicit test to None I think.

Reply:
Actually now I am wondering more about this. Since x can be Any - not necessarily Iterable - it seems there is something more fundamental that needs addressing here. Why is x treated different than say fval - they are both Any